packages/@aws-cdk/assertions/lib/template.ts

```diff
@@ -109,10 +109,11 @@ export class Template {
    * Assert that an Output with the given properties exists in the CloudFormation template.
    * By default, performs partial matching on the resource, via the `Match.objectLike()`.
    * To configure different behavour, use other matchers in the `Match` class.
+   * @param outputName the name of the output.
    * @param props the output as should be expected in the template.
    */
-  public hasOutput(props: any): void {
-    const matchError = hasOutput(this.inspector, props);
+  public hasOutput(outputName: string, props: any): void {
```

Comment:
We should probably also change `findOutputs()`.

Reply:
It seems like for `findOutputs`, we should be less stringent about the `outputName` and still allow for users to search for some `value` among all outputs like so:

```ts
test('matching', () => {
      const stack = new Stack();
      new CfnOutput(stack, 'Foo', {
        value: 'Fred',
        description: 'FooFred',
      });
      new CfnOutput(stack, 'Bar', {
        value: 'Fred',
        description: 'BarFred',
      });
      new CfnOutput(stack, 'Baz', {
        value: 'Waldo',
        description: 'BazWaldo',
      });

      const inspect = Template.fromStack(stack);
      const result = inspect.findOutputs({ Value: 'Fred' });
      expect(result).toEqual([
        { Value: 'Fred', Description: 'FooFred' },
        { Value: 'Fred', Description: 'BarFred' },
      ]);
    });
```

With that in mind, do you think we should:

a) still require an `outputName` in `findOutputs` (what I read that you are suggesting in the above comment)
b) make `outputName` optional
c) figure out some way to return the `outputName` so the user has the full picture.

If it were up to me I think that `c)` is the best option though it would require us to make a change to `matchSection()` so we return the `logicalId` along with the rest of the output. What do you think?

Reply:
I like your suggestion of making `outputName` optional and returning the `logicalId`. How about we split this into two changes? In this PR, implement (b) and in a subsequent PR, implement (c) but for all `find*()` APIs.

Reply:
It seems like making `outputName` optional is problematic in its own right. We get `public findOutputs(outputName?: string, props?: string)` which doesn't really work because of the two optional parameters. The user experience for not wanting to specify an `outputName` would be `findOutputs(undefined, MyProps)`. I feel like that is unacceptable, but I'm open to input.

Otherwise, the three options I can think of are overloading the method (but typescript barely supports this and I have no reference that this happens anywhere in the CDK) or refactoring to a method signature of `findOutputs(options: findOptions)`. I am also sympathetic towards having all the `find` and `has` methods similar in signature, so perhaps it is time to start thinking about all of these methods taking in a `findXXXOptions` or `hasXXXOptions` instead.

My last option is the one I like best and the one I am currently implementing. I think we should require `outputName` and clearly document that a particular token (`'*'`) represents all outputs. Then, the user experience for not wanting to specify an outputName would be `findOutputs('*', MyProps)` which I think makes more sense than `undefined`.